content/guides/end-to-end-testing/testing-your-app.md

```diff
@@ -410,111 +410,87 @@ You'll likely also want to test your login UI for:
-
-</Alert>
-
-Let's investigate and tease apart why.
-
-#### Bypassing your UI
-
-When you're writing tests for a very **specific feature**, you _should_ use your
-UI to test it.
+```js
+// In cypress/support/commands.js
 
-But when you're testing _another area of the system_ that relies on a state from
-a previous feature: **do not use your UI to set up this state**.
+Cypress.Commands.add('login', (username, password) => {
+  cy.visit('/login')
 
-Here's a more robust example:
+  cy.get('input[name=username]').type(username)
 
-Imagine you're testing the functionality of a **Shopping Cart**. To test this,
-you need the ability to add products to that cart. Well where do the products
-come from? Should you use your UI to login to the admin area, and then create
-all of the products including their descriptions, categories, and images? Once
-that's done should you then visit each product and add each one to the shopping
-cart?
+  // {enter} causes the form to submit
+  cy.get('input[name=password]').type(`${password}{enter}`)
 
-No. You shouldn't do that.
+  // we should be redirected to /dashboard
+  cy.url().should('include', '/dashboard')
 
-<Alert type="warning">
-
-<strong class="alert-header">Anti-Pattern</strong>
+  // our auth cookie should be present
+  cy.getCookie('your-session-cookie').should('exist')
 
-Don't use your UI to build up state! It's enormously slow, cumbersome, and
-unnecessary.
+  // UI should reflect this user being logged in
+  cy.get('h1').should('contain', username)
+})
 
-Read about [best practices](/guides/references/best-practices) here.
+// In your spec file
 
-</Alert>
+it('does something on a secured page', function () {
+  const { username, password } = this.currentUser
+  cy.login(username, password)
 
-Using your UI to **log in** is the _exact same scenario_ as what we described
-previously. Logging in is a prerequisite of state that comes before all of your
-other tests.
+  // ...rest of test
+})
+```
 
-Because Cypress isn't Selenium, we can actually take a huge shortcut here and
-skip needing to use our UI by using [`cy.request()`](/api/commands/request).
+#### Improving performance
 
-Because [`cy.request()`](/api/commands/request) automatically gets and sets
-cookies under the hood, we can actually use it to build up state without using
-your browser's UI, yet still have it perform exactly as if it came from the
-browser!
+You're probably wondering what happened to our advice about logging in "only
+once". The custom command above will work just fine for testing your secured
+pages, but if you have more than a handful of tests, logging in before every
+test is going to add a lot of running time to your suite.
 
-Let's revisit the example from above but assume we're testing some other part of
-the system.
+Luckily, Cypress provides the [`cy.session()` command](/api/commands/session), a
+powerful performance tool that lets you cache user session data and reuse it for
```

Comment:
```suggestion
powerful performance tool that lets you cache the browser context associated to your user's and reuse it for
```
session data is pretty overloaded and means different things to different users. Being explicit and saying cookies, local storage and session storage might be better.